app/queries/drug_stocks_query.rb

```diff
@@ -0,0 +1,103 @@
+class DrugStocksQuery
+  CACHE_VERSION = 1
+
+  def initialize(facilities:, for_end_of_month:)
+    @facilities = facilities
+    @for_end_of_month = for_end_of_month
+    # assuming that all facilities on the page have the same protocol
+    @protocol = @facilities.first.protocol
+    @state = @facilities.first.state
+  end
+
+  def protocol_drugs_by_category
+    @protocol_drugs_by_category ||= @protocol.protocol_drugs
+      .where(stock_tracked: true)
+      .order(:name, :dosage)
+      .sort_by(&:sort_key)
+      .group_by(&:drug_category)
+      .sort_by { |(drug_category, _)| drug_category }
+      .to_h
+  end
+
+  def cache_key
+    [
+      self.class.name,
+      @facilities.map(&:id).sort,
+      @for_end_of_month,
+      @protocol.id,
+      @state,
+      CACHE_VERSION
+    ].join("/")
+  end
+
+  def call
+    Rails.cache.fetch(cache_key,
+      expires_in: ENV.fetch("ANALYTICS_DASHBOARD_CACHE_TTL"),
+      force: RequestStore.store[:force_cache]) do
+      {all: totals,
+       facilities: report_for_facilities,
+       last_updated_at: Time.now}
+    end
+  end
+
+  def patient_counts
+    @patient_counts ||= Patient.where(assigned_facility_id: @facilities).group(:assigned_facility).count
+  end
+
+  def drug_stocks
+    @drug_stocks ||= DrugStock.latest_for_facilities(@facilities, @for_end_of_month).group_by(&:facility_id)
+  end
+
+  def report_for_facilities
+    @report_for_facilities ||= @facilities.each_with_object({}) { |facility, report|
+      report[facility.id] = drug_stock_for_facility(facility, drug_stocks[facility.id])
+    }
+  end
+
+  def drug_stock_for_facility(facility, facility_drug_stocks)
+    patient_count = patient_counts[facility] || 0
+    facility_report = {facility: facility, patient_count: patient_count}
+    return facility_report if facility_drug_stocks.nil?
+
+    drug_stocks = facility_drug_stocks.group_by { |drug_stock| drug_stock.protocol_drug.drug_category }
+    protocol_drugs_by_category.each do |(drug_category, _protocol_drugs)|
+      stocks = stocks_by_rxnorm_code(drug_stocks[drug_category])
+      patient_days = Reports::PatientDaysCalculation.new(@state, @protocol, drug_category, stocks, patient_count).calculate
+      next if patient_days.nil?
+      facility_report[drug_category] = patient_days.merge(drug_stocks: drug_stocks_by_drug_id(drug_stocks[drug_category]))
+    end
+    facility_report
+  end
```

Comment:
This method is pretty dense, I'm having a tough time understanding what's going on here. Can you post an example payload? I'm curious if we can work backwards from there to craft a more readable way to express these operations.

Reply:
There are a couple of examples [in the spec](https://github.com/simpledotorg/simple-server/pull/2012/files/f6879780cc577fb49e4a5b74b17367da74c5da77#diff-e82c91761e53671a526dd257d939fad4211ce7f669ec69a2a02cdeebdbf7d1fcR36) that would help here I think.

Here's one way to think about this. There are two hierarchies:
1. Protocol -> Protocol Drug Categories -> Protocol Drugs (columns)
2. Facility -> Protocol Drug Categories -> Drug Stocks [and patient days] (rows)

We need to traverse through both these for every row, since there might be protocol drugs that don't have drug stocks in that facility, etc.

I do agree that the method is dense, however. Totally open to ideas. We could extract a method for each drug category so it's easier to read perhaps.

Reply:
Extracted in 9877b110588afbc3cabb398af064bf59ab7b5967. We feel this reads easier. We've swept some of the complexity under the carpet of another function, but this is perhaps good enough for now.

Given that the data structure here is quite nested, we could probably use a better abstraction that would return normalised data for each row. We can go there the next time we visit this, or when we look at the consumption report perhaps.

-- @vkrmis and @ssrihari

Reply:
Surprisingly, the method extraction actually cleans up the readability quite nicely! ✨ The nested structure of facility -> drug category -> stocks is more evident now. Nice.